Classifier: Custom Video Controls

goplayoutside3 opened the pull request:
~~Blocked because this PR is based on #3645.~~

## Package
lib-classifier

## Linked Issue and/or Talk Post
#3612 and #3613 

## Describe your changes

This is a complete rewrite of the VideoController component. The controls are designed to sit below a html video component so that future drawing tools can be positioned over a video subject. Not all video subject projects will use drawing tools, but for consistency these custom video controls will always display.

### File Re-organization

I consolidated all aspects of the controls into one component: VideoController. It lives in a /components folder in SingleVideoViewer. I also moved the `react-player` component into SingleVideoViewerContainer for readability. All of its handler functions are defined in SingleVideoViewerContainer.

### Controls Functionality

I improved the video scrubbing by memoizing the `react-player` so it's `onProgress` and `setTimeStamp` functions don't call at the same time.

I edited `formatTimeStamp()` to display timestamps in the same units as popular video players such as native html video controls, youtube, vimeo, etc. (i.e. 0:20 means 20 seconds).

I added the ability to change volume and display a video subject in fullscreen.

### Controls Styling

Screenshots for review: 
Note that there is a 1px white border on top of the video controls because the Woodpecker Cavity Cam subjects have black borders, and I want the video controls' dimensions to be obvious. The white border only appears in the screenshots, not in the code.

![woodpecker desktop](https://user-images.githubusercontent.com/23665803/191377047-7ba2c01c-c792-43c7-b1e7-f4161e02b818.jpg)

![woodpecker smaller screen](https://user-images.githubusercontent.com/23665803/191377197-ef5ca0d5-c493-4aab-8df3-94744d12ec5a.jpg)

![woodpecker volume](https://user-images.githubusercontent.com/23665803/191377205-d61031c6-7e2c-441e-a2c3-fec75428708d.jpg)

Should the controls be re-arranged on a mobile view?
![woodpecker mobile](https://user-images.githubusercontent.com/23665803/191377240-8254d51a-4344-487e-abf8-dfa2d800631c.jpg)

### Documentation and Tests

I added documentation in a README file for VideoController and SingleVideoViewerContainer. I added unit tests using React Testing Library for VideoController.

## How to Review
 This PR is part of an effort to improve the video subject viewer code for migrating [Woodpecker Cavity Cam](https://www.zooniverse.org/projects/elwest/woodpecker-cavity-cam) to FEM, so I recommend that project for reviewing functionality.

Things to check a volunteer can do:
- Play / Pause the video
- See the timestamp and duration of video subject
- Scrub the video with the slider
- Change the playback rate
- Change the volume
- View the video subject in fullscreen

Video Controller also appears in SingleVideoViewer's storybook.

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

seanmiller26 reviewed (commented):
The controls look great! As far as the mobile view question, yes the rearrangement looks great. I'm wondering if there's a way to have it closer to the subject image? Or is it in a fixed position? (see screenshot) I know it would also depend on the size of the subject images.

I really want to address the margin width at some point in the near future. We've had at least one complaint about this issue. There is too much margin space (including the space b/w the subject image and task area) and this takes valuable space away from both areas. But this can wait until we fully address other responsive issues.
<img width="481" alt="Screen Shot 2022-09-22 at 12 08 06 PM" src="https://user-images.githubusercontent.com/89043489/191811535-e3cc8f7d-83d3-44e8-867d-94812d0a3029.png">

eatyourgreens commented:
Not a blocker for this PR, but the zoom, pan, rotate and reset buttons in the image toolbar don't make much sense for a video subject.

goplayoutside3 commented:
> I'm wondering if there's a way to have it closer to the subject image? Or is it in a fixed position? (see screenshot)

@seanmiller26  The "Subject Info" and "Add to" buttons are dependent on the height of the Image Toolbar regardless of screen size. I think it'd be best to write a separate Issue for the layout. For instance, this is HMS NHS on mobile: 

![hms nhs](https://user-images.githubusercontent.com/23665803/191815105-7c778c78-f276-4290-a994-8dd2e0f92a9e.jpg)

> Not a blocker for this PR, but the zoom, pan, rotate and reset buttons in the image toolbar don't make much sense for a video subject.

@eatyourgreens  Agreed! I think some buttons in the image toolbar should be removed or disabled for projects with video subjects in a follow-up PR.

eatyourgreens commented:
When we were working on the TESS light curve viewer, we did add a way to disable at least some of those buttons, but I can't remember how. It's probably set in the `SubjectViewer` model.

goplayoutside3 commented:
> @seanmiller26 The "Subject Info" and "Add to" buttons are dependent on the height of the Image Toolbar regardless of screen size. 

For clarity, I did not touch those buttons in this PR, I only styled the controls displayed in black below a video subject.

seanmiller26 commented:
@goplayoutside3 Sorry, not trying to derail this PR. I thought that's what you were asking about:

> Should the controls be re-arranged on a mobile view?

Thanks for the clarification, though. And yes, definitely agree with showing less toolbar controls depending on the subject type

shaunanoordin commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/helpers/getFixedNumber/getFixedNumber.js:
```diff
@@ -1,3 +1,6 @@
 export default function getFixedNumber(number, digits) {
+  if (!digits) {
+    console.log('getFixedNumber: Missing digits variable')
+  }
```
[minor][rambling] It's good that we're checking for input here, but if we don't have `digits` (i.e. undefined), won't this function still return `NaN`? Hmm, but that said, I don't know what's the correct "fallback" return value should be if `digits` is undefined... eh, NaN is fine.

goplayoutside3 replied:
My thoughts too 😆 For now, I wanted to return a fallback so the subject viewer wouldn't crash.

shaunanoordin commented on packages/lib-classifier/src/helpers/formatTimeStamp/README.md:
```diff
@@ -1,19 +1,19 @@
 E.g. `0.287` or `0.962`
 
-**duration:** Optional. Type: number. This is the total length of the video. This must be a number with 3 decimal places.
+**duration:** Optional. Type: number. This is the total length of the video. This must be a number.
```
[documentation] This is in seconds, right?

```suggestion
**duration:** Optional. Type: number. This is the total length of the video, in seconds. This must be a number.
```

goplayoutside3 replied:
You're right!

shaunanoordin commented on packages/lib-classifier/src/helpers/formatTimeStamp/README.md:
```diff
@@ -1,19 +1,19 @@
+A helper function to convert a number in seconds to a human readable string `mm:ss`. We're assuming the timestamp of a video subject is always less than 1 hour.
 
-The Zooniverse video player has a slider who's values are 0.0 to 1.0. As a video plays, the handle will progress from 0.0 (start of the video) to 1.0 (end of the video). This function converts the decimal number to a human readable timestamp and is used in the video controller and sent as a property for temporal drawing tools.
+The Zooniverse video player has a slider who's percentage `timeStamp` values are 0 to 1. As a video subject plays, the slider input will progress from 0 (start of the video) to 1 (end of the video). This function converts the decimal number to a human readable timestamp. The returned value is used in the VideoController and sent to the store as a property for temporal drawing tools.
```
[minor]

```suggestion
The Zooniverse video player has a slider whose percentage `timeStamp` values are 0 to 1. As a video subject plays, the slider input will progress from 0 (start of the video) to 1 (end of the video). This function converts the decimal number to a human readable timestamp. The returned value is used in the VideoController and sent to the store as a property for temporal drawing tools.
```

shaunanoordin reviewed (commented):
## PR Review (initial)

Package: `lib-classifier`

This PR rewrites and updates various parts of the Single Video Viewer

- Video Controller component has been completely rewritten.
  - As noted in the new README, the Video Controller is separate from the `react-player`'s built-in controls because we need to place a "drawing canvas" directly on top of the _full visual space_ of the react-player, and the built-in controls would obscure that. 
- Documentation: added for Video Controller, updated elsewhere.
- Tests: updated in various places

The _functionality_ of the Single Video Viewer _is intended_ to remain the same, if I understand this PR correctly.

Code read looks good, though I've added a bunch of comments, including:
- notes on the `playbackRate` going from a number like `1` to a string like `"1x"`
- some suggestions for [documentation] updates, to add clarity to some bits
- some other [minor] suggestions.

No PR blockers from me, at least at this code read stage.

### Status

Code read complete, now moving on to manual testing. brb.

shaunanoordin reviewed (approved):
## PR Review (Update)

Follows [initial review](https://github.com/zooniverse/front-end-monorepo/pull/3684#pullrequestreview-1127342617)

The functionality of the Single Video Viewer and, more importantly, the rewritten Video Controller mostly checks out OK. 👍 I do have some comments, as usual:

- Basic functionality: 👍 
  - I can view a video subject.
  - I can click on the play/pause button to play/pause the video.
  - I can view the current time and full duration of the video, e.g. `0:03 / 0:10`
  - I can click on the video-time-scrubber-bar (what IS this called?) OR drag the time-position-handle-thing (what IS this also called?) to change the currently viewed time of the video.
    - ⚠️ NOTE: there is some lag between setting the time, and the video jumping to that time.
  - I can click on the the playback rate to display speed options, then select one to change the speed the video is played
  - I can click on the volume button to display and adjust the video volume.
  - I can click on the fullscreen button to enter fullscreen mode. (I can also exit this mode.)
- Accessibility (keyboard controls):
  - I can use a keyboard to navigate through the interactable video controller elements (play button, time-scrubber-bar, playback rate dropdown/dropup, volume controls, fullscreen button)
  - I can press Spacebar or Enter on the play/pause button to play/pause the video.
    - ⚠️ NOTE: there seems to be a small delay on the time-scrubber-bar thing. Like if I press pause at 0:02, the video correctly pauses, but the time-scrubber-bar thing keeps moving forward for a split-second as if it has some momentum then stops at 0:03. It doesn't make the video controls unusable, but it's weird.
  - ❌ I **can't** use the left/right arrow keys, on the time-scrubber-bar thing, to step back/step forward through the video time. Essentially, I can't use the keyboard to scrub.
  - ❌ While I'm on the time-scrubber-bar thing, pressing enter/spacebar does **not** pause/play the video. (This is standard behaviour for, say, a YouTube player)
  - I can use the keyboard to view & select the playback speed
  - ❌ I can use the keyboard to open the volume menu, but I **can't** use the up/down keys to adjust the volume on an open volume menu.
  - I can press Spacebar or Enter on the Fullscreen button to enable fullscreen mode, then press Esc to escape fullscreen mode.
- Accessibility (labelling/screen readers):
  - ❔ I have no useful comments here, sorry - might be worth seeking an additional opinion on this.
- Advanced?
  - ❓ When the video is in fullscreen, the video controller appears OVER (overlaps with) the video.
    - I'm not sure if fullscreen would affect any of the drawing tasks, but that's probably a problem for another PR.
    - I'm also not sure if this problem is specific to the size of the video + my full screen size.

### Testing

Tested on localhost with macOS + Chrome 105, with [Woodpecker Cavity Cam](https://local.zooniverse.org:8080/?env=production&project=elwest%2Fwoodpecker-cavity-cam&workflow=16998) (production) as the primary test project.

Testing steps: see basic functionality/accessibility/etc above.

### Status

Basic functionality checks out in most cases, so this PR gets a 👍 from me.

One major thing to note is that the keyboard controls are far more limited than what you'd expect from a standard video player (see all the ❌ icons above) - but I think these keyboard controls can be addressed in a separate PR. I can write the missing functionality up as an issue, if that helps?

eatyourgreens commented:
That’s odd. The [range input](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range)  responds to the cursor keys by default, when it has focus. 

I’m not sure that we should merge PRs that have an error like this. Thoughts?

eatyourgreens commented:
When a range opens in a popup, like the volume control, you probably need to call `volumeControl.focus()` to focus it. Grommet might not do that for you. It’s also good practice to focus the first focusable element when a popup opens, and check that focus returns to the original button when the popup closes.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleVideoViewer/components/VideoController/README.md:
```diff
@@ -0,0 +1,27 @@
+- `handleVolumeOpen`: (func) Function passed from SingleVideoViewerContainer and called when volume icon is clicked.
+- `onPlayPause`: (func) Function passed from SingleVideoViewerContainer and called when player is played or paused.
+- `onSpeedChange`: (func) Function passed from SingleVideoViewerContainer and called when the playback rate Select component is changed.
+- `onSliderChange`: (func) Function passed from SingleVideoViewerContainer and called when Slider input is changed.
```
MDN documents change events for range inputs. Would it be helpful to link to that? I’m not sure if that’s what you’re asking @shaunanoordin? I read your comment as ‘explain how change events work.’

goplayoutside3 replied:
I think I lean toward Jim's suggestion. For instance, the variable passed to `onSliderChange()` is distinctly labeled as `e.target.value` in SingleVideoViewerContainer.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleVideoViewer/SingleVideoViewerContainer.js:
```diff
@@ -113,25 +112,75 @@ function SingleVideoViewerContainer({
-    setIsSeeking(false)
+  const handleSliderChange = e => {
+    const newTimeStamp = e.target.value
+    playerRef?.current.seekTo(newTimeStamp, 'seconds')
```
`playerRef.current` will be null when the component mounts. I think the ref will be set after the first render. Probably safe here, if this is a handler that’s called once the player DOM exists.

goplayoutside3 replied:
`handleSliderChange()` is used in VideoController. VideoController renders independently of `playerRef`, so I'll add `current?` here.

eatyourgreens commented:
Regarding poor performance of the range input for the scrubber, have you considered using an uncontrolled input rather than a controlled input. There’s a lot of internal component state at the moment, which would make the controls laggy. The current play time can be read from the video, via the Media API. Updating it via `setState` adds extra render cycles, which I think is what leads to the lag between pausing the video and updating the progress bar.

shaunanoordin commented:
> I’m not sure that we should merge PRs that have an error like this. Thoughts?

👌 I'm OK to have this PR merged first, and then have the keyboard issues fixed in a separate PR. My rationale: (1) this PR doesn't break anything that exists, and (2) a second smaller PR would be very helpful in understanding what was done to solve the keyboard issues for the video player.

To expand on (1), this is how the current video controls work on `master` (ad1d18b):
- When focused on the time-scrub-bar,
  - ...you _can_ use keyboard left/right to adjust current time, but at a _per millisecond_ rate. (Chrome 105, FF 105) This makes it practically unusable, as it takes approx 5-6 minutes to rewind the video by 1 second.
  - ...you _still can't_ use spacebar to pause/play the video
- There's no explicit volume control.

(Side note: the poor/laggy performance of the video control when you scrub through the time bar also exists on the old version of the video player.)

eatyourgreens commented:
The scrubber in the current player can be controlled from the keyboard, so that would be a breaking change here.

eatyourgreens commented:
Actually I might be thinking of the video player in live projects. That should be our reference for breaking changes.

eatyourgreens commented:
Also on [Woodpecker Cavity Cam](https://www.zooniverse.org/projects/elwest/woodpecker-cavity-cam/classify), in Firefox, I can mute/unmute the volume control with the spacebar. When volume is unmuted, I can operate the volume slider with the cursor keys.

https://user-images.githubusercontent.com/59547/193597300-fe287725-3b44-41da-81df-6109cf5484c8.mov

goplayoutside3 commented:
@eatyourgreens @seanmiller @shaunanoordin how do you feel about the following scenario? I'd like to propose we do the following to simplify all the discussion initiated by this PR.

Our goal during FEM migration of the first few pathfinder projects is to migrate classification in an uninterrupted manner from PFE to FEM. Woodpecker Cavity Cam is a survey task project with video subjects _without_ drawing tools. In PFE, the subjects are displayed with native browser controls that automatically show with an HTML video element.

Proposition: I modify this PR to show FEM custom video controls ONLY if a project has drawing tools, and Woodpecker Cavity Cam in FEM would have the same UX as it currently does in PFE. Then, I create Github Issues for each point you've commented in this PR as future work toward FEM video subjects _with_ drawing tools and therefore with custom video controls. 

While developing the video subject viewer specifically for an un-launched video + drawing project, I think we've missed the opportunity to build the simple video subject viewer based on PFE.

Thoughts?

eatyourgreens commented:
That sounds like a really good idea to me. I'd definitely avoid the extra work of creating controls that already exist in the browser. :+1:

Maybe have two video player components, rather than one player with branching logic and optional custom controls?

shaunanoordin commented:
> Proposition: I modify this PR to show FEM custom video controls ONLY if a project has drawing tools, and Woodpecker Cavity Cam in FEM would have the same UX as it currently does in PFE. Then, I create Github Issues for each point you've commented in this PR as future work toward FEM video subjects with drawing tools and therefore with custom video controls.

That gets a solid thumbs up from me. 👍

eatyourgreens commented:
One thought: I think we may still need custom controls for video speed. Speed control might be available on Android and iOS but I'm not sure about other operating systems.

seanmiller26 commented:
> Proposition: I modify this PR to show FEM custom video controls ONLY if a project has drawing tools, and Woodpecker Cavity Cam in FEM would have the same UX as it currently does in PFE. Then, I create Github Issues for each point you've commented in this PR as future work toward FEM video subjects with drawing tools and therefore with custom video controls.

If this means far less work, then I'm all for it. I'd like to keep in mind uniformity and conciseness as an end gaol, though. We don't want to have multiple different players etc for slightly different subject tasks - which could lead to confusion from users.
So eventually transitioning to both these tasks having the same video controls would be ideal, even if that means waiting until we have the time to do it properly.

goplayoutside3 commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleVideoViewer/README.md:
```diff
@@ -24,12 +24,14 @@ Props
+- `fullscreen`: (boolean) Whether or not the video is displayed fullscreen.
 - `isPlaying`: (boolean) Whether or not the video subject is playing.
-- `isSeeking`: (boolean) Whether or not a user is interacting with the VideoController > Slider.
 - `playbackRate`: (number) 1, 0.5, or 0.25 ratio determines the speed of video playback.
```
Totally, I'll update the README.

goplayoutside3 commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleVideoViewer/SingleVideoViewerContainer.js:
```diff
@@ -40,12 +40,14 @@ function SingleVideoViewerContainer({
   const [isPlaying, setIsPlaying] = useState(false)
-  const [isSeeking, setIsSeeking] = useState(false)
-  const [playbackRate, setPlaybackRate] = useState(1)
+  const [playbackRate, setPlaybackRate] = useState('1x')
```
I agree. At the time, I did this so the VIdeoController's speed control's labels will match that of PFE. I like your suggestion to store a pure value in state and then render a string label such as 0.5x. I'll leave it as is, and create a to-do in the Video Annotator project board to refactor while I add simple speed control buttons intended to display with the native browser video controls in non-drawing-tools projects.

goplayoutside3 commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleVideoViewer/components/VideoController/VideoController.js:
```diff
@@ -0,0 +1,192 @@
+        <Box direction='row' gap='small'>
+          {/* Rate */}
+          <Select
+            a11yTitle={t('SubjectViewer.VideoController.playbackSpeed')}
```
Good catch! I haven't added the new translation keys for custom video controls to dictionaries other than English yet, so I'll correct this naming for clarity.

goplayoutside3 commented:
> ❓ When the video is in fullscreen, the video controller appears OVER (overlaps with) the video.
I'm not sure if fullscreen would affect any of the drawing tasks, but that's probably a problem for another PR.
I'm also not sure if this problem is specific to the size of the video + my full screen size.

Re: @shaunanoordin I don't think fullscreen works with drawing tools, so I've added a check that says to only display the fullscreen button when drawing tools are disabled.

> One thought: I think we may still need custom controls for video speed. Speed control might be available on Android and iOS but I'm not sure about other operating systems.

Re: @eatyourgreens Playback speed is available in native desktop browser control. PFE does display custom buttons though, so I wrote a to-do Issue to add them to FEM as well (Issue #3758).

### Update

I added an ADR to this PR describing the decision to use native browser video controls to aid in PFE to FEM migration. I also added stories and unit tests for the drawing tools disabled versus enabled UI. Accessibility issues are written in Issue #3759. I'm going to merge this PR this evening.
